### PR TITLE
feat: Add Rust SDK

### PR DESCRIPTION
**Pull request by jeremyhager**
### Motivation
The SDK for Argo Workflows provides 2 official packages: Python and Java. However this is lacking for other languages such as, and in particular for this PR, Rust. This PR adds Rust into the core Argo Workflows repo.

### Modifications

- Add SDK for rust under `sdks/rust`
  - `Makefile` for generating, "installing", and testing
  - `README.md` for getting started
  - `.gitignore` as necessary for files and folders
- Modify `.github/ci-build.yaml`
  - Add test for Rust SDK under `e2e-tests`
  - Add `actions-rs` for Rust `matrix-test`
  - Update `make cli STATIC_FILES=false` with Rust matrix test
  - Update `make start PROFILE` with Rust matrix test
  - Update `Start controller/API` with Rust matrix test
- Modify `.github/sdks.yaml`
  - Update strategy to include `rust`
  - Update to include `CARGO_REGISTRY_TOKEN`
- Modify `Makefile`
  - Add `make --directory sdks/rust generate` to `codegen` command
 
### Verification

- Run k3d local cluster and `make test` rust sdk against local
- Run README.md tests within `sdks/rust` to confirm examples work
- Run `make pre-commit -B`
- Run draft PR to check changes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -256,6 +256,7 @@ codegen: types swagger manifests $(GOPATH)/bin/mockery docs/fields.md docs/cli/a
 	go generate ./...
 	make --directory sdks/java generate
 	make --directory sdks/python generate
+	make --directory sdks/rust generate
```
Comment (agilgur5) [inline Makefile:259]: Will want to add the generated files to the `.gitattributes`, similar to the [existing generated SDKs](https://github.com/argoproj/argo-workflows/blob/82310dd459aa080754169d6a1667d30d9b7c75bf/.gitattributes#L1)

**Comment by isubasinghe** [comment]
@jeremyhager are you attempting to use argo workflows via Rust? Just wonder what the interest is for adding a Rust sdk. 

Even though this is sorta free code (given that its generated), I still have some concerns about taking on maintenance of it. 
The reason being is that for example the python sdk has been reported to be buggy, not sure on this but I think there are some talks about replacing the python sdk with hera. 

What I am trying to get at here is the fact that I would very much like to avoid maintenance of a rust sdk if there isn't much community interest in it. 

If you can't tell already, I love Rust, I just am unconvinced that a Rust sdk is needed. 

That being said, I do appreciate your time for working on this.

**Comment by jeremyhager** [comment]
@isubasinghe Indeed, was thinking to create a Rust SDK for Argo Workflows. I also asked on the Slack channel, just to gauge interest.

Some of the concerns I have with the auto-generated code is that the testing requires a few helper functions that would do better just to be included - for example a `::new()` for some structs, but who knows if that itself would be overall helpful. I wasn't aware that the Python SDK was buggy, or at least there has been talk to replace it with hera. Because it's auto-generated, there aren't a lot of options for improving the existing code without adjusting the generation files. Even then, the engine itself may need to be updated.

In the end, it seems there's some community interest, but I will freely admit there doesn't seem to be a lot - you could be right, an sdk may be a bit overkill.

Right now I'm just working through getting the tests to run, but it's been a minute since I've had the time to work on it.

**Comment by agilgur5** [comment]
Relevant note: It was discussed in [this week's contributor meeting](https://docs.google.com/document/d/1tznN5LB-KrNkC6dmeIzpKfuyvZWgCJpRuS-E84zseC8/edit#heading=h.dog2y3c5ev54) that new SDKs might be better placed into [`argoproj-labs`](https://github.com/argoproj-labs) until they are stable & well maintained.
I wasn't present for that meeting, but that sounds like a good proposal to me.

**Comment by danmx** [comment]
Any updates on the decision?

**Comment by terrytangyuan** [comment]
@jeremyhager We can create a repo in argoproj-labs for new language SDKs. WDYT?

**Comment by jeremyhager** [comment]
@terrytangyuan I think that's a good idea. I can close this out and work on a PR on the argoproj-labs repo

**Comment by danmx** [comment]
When do you plan to publish the repo? I don't see it in the labs org

**Pull request closed without merging**